File: csla20cs/Csla/Web/Design/ObjectFieldInfo.cs
using System;
using System.Data;
using System.Web.UI;
using System.Web.UI.Design;
using System.ComponentModel;
using System.Reflection;
using System.Collections;
using System.Collections.Generic;

namespace Csla.Web.Design
{
  /// <summary>
  /// Contains schema information for a single
  /// object property.
  /// </summary>
  public class ObjectFieldInfo : IDataSourceFieldSchema
  {
    private PropertyDescriptor _field;
    private bool _primaryKey;
    private bool _isIdentity;
    private bool _isNullable;
    private int _length;

    /// <summary>
    /// Creates an instance of the object.
    /// </summary>
    /// <param name="field">The PropertyInfo object
    /// describing the property.</param>
    public ObjectFieldInfo(PropertyDescriptor field)
    {
      _field = field;
      GetDataObjectAttributes();
    }

    private void GetDataObjectAttributes()
    {
      DataObjectFieldAttribute attribute =
        (DataObjectFieldAttribute)
        _field.Attributes[typeof(DataObjectFieldAttribute)];
      if (attribute != null)
      {
        _primaryKey = attribute.PrimaryKey;
        _isIdentity = attribute.IsIdentity;
        _isNullable = attribute.IsNullable;
        _length = attribute.Length;
      }
    }

    /// <summary>
    /// Gets the data type of the property.
    /// </summary>
    public Type DataType
    {
      get
      {
        return Utilities.GetPropertyType(
          _field.PropertyType);
      }
    }

    /// <summary>
    /// Gets a value indicating whether this property
    /// is an identity key for the object.
    /// </summary>
    /// <remarks>
    /// Returns the optional value provided through
    /// the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    /// attribute on the property.
    /// </remarks>
    public bool Identity
    {
      get { return _isIdentity; }
    }

    /// <summary>
    /// Gets a value indicating whether this property
    /// is readonly.
    /// </summary>
    public bool IsReadOnly
    {
      get { return !_field.IsReadOnly; }
    }

    /// <summary>
    /// Gets a value indicating whether this property
    /// must contain a unique value.
    /// </summary>
    /// <returns>
    /// Always returns True if the property
    /// is marked as a primary key, otherwise
    /// returns False.
    /// </returns>
    public bool IsUnique
    {
      get { return _primaryKey; }
    }

    /// <summary>
    /// Gets the length of the property value.
    /// </summary>
    /// <remarks>
    /// Returns the optional value provided through
    /// the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    /// attribute on the property.
    /// </remarks>
    public int Length
    {
      get { return _length; }
    }

    /// <summary>
    /// Gets the property name.
    /// </summary>
    public string Name
    {
      get { return _field.Name; }
    }

    /// <summary>
    /// Gets a value indicating whether the property
    /// is nullable
    /// </summary>
    /// <remarks>
    /// Returns True for reference types, and for
    /// value types wrapped in the Nullable generic.
    /// The result can also be set to True through
    /// the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    /// attribute on the property.
    /// </remarks>
    public bool Nullable
    {
      get
      {
        Type t = _field.PropertyType;
        if (!t.IsValueType || _isNullable)
          return true;
        if (t.IsGenericType)
          return (t.GetGenericTypeDefinition() == typeof(Nullable<>));
        return false;
      }
    }

    /// <summary>
    /// Gets the property's numeric precision.
    /// </summary>
    /// <returns>Always returns -1.</returns>
    public int Precision
    {
      get { return -1; }
    }

    /// <summary>
    /// Gets a value indicating whether the property
    /// is a primary key value.
    /// </summary>
    /// <remarks>
    /// Returns the optional value provided through
    /// the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    /// attribute on the property.
    /// </remarks>
    public bool PrimaryKey
    {
      get { return _primaryKey; }
    }

    /// <summary>
    /// Gets the property's scale.
    /// </summary>
    /// <returns>Always returns -1.</returns>
    public int Scale
    {
      get { return -1; }
    }
  }
}

File: cslavb/Csla/Web/Design/ObjectFieldInfo.vb
Imports System.Web.UI
Imports System.Web.UI.Design
Imports System.ComponentModel
Imports System.Reflection

Namespace Web.Design

  ''' <summary>
  ''' Contains schema information for a single
  ''' object property.
  ''' </summary>
  <Serializable()> _
  Public Class ObjectFieldInfo

    Implements System.Web.UI.Design.IDataSourceFieldSchema

    Private _dataType As Type
    Private _primaryKey As Boolean
    Private _isIdentity As Boolean
    Private _isNullable As Boolean
    Private _length As Integer
    Private _isReadOnly As Boolean
    Private _name As String
    Private _nullable As Boolean

    ''' <summary>
    ''' Creates an instance of the object.
    ''' </summary>
    ''' <param name="field">The PropertyInfo object
    ''' describing the property.</param>
    Public Sub New(ByVal field As PropertyDescriptor)

      Dim attribute As DataObjectFieldAttribute = _
          CType(field.Attributes.Item(GetType(DataObjectFieldAttribute)),  _
          DataObjectFieldAttribute)
      If (Not attribute Is Nothing) Then
        With attribute
          _primaryKey = .PrimaryKey
          _isIdentity = .IsIdentity
          _isNullable = .IsNullable
          _length = .Length
        End With
      End If
      _dataType = Utilities.GetPropertyType(field.PropertyType)
      _isReadOnly = field.IsReadOnly
      _name = field.Name

      Dim t As Type = field.PropertyType
      If Not t.IsValueType OrElse _isNullable Then
        _nullable = True

      ElseIf t.IsGenericType Then
        _nullable = (t.GetGenericTypeDefinition Is GetType(Nullable(Of )))

      Else
        _nullable = False
      End If

    End Sub

    ''' <summary>
    ''' Gets the data type of the property.
    ''' </summary>
    Public ReadOnly Property DataType() As System.Type _
      Implements System.Web.UI.Design.IDataSourceFieldSchema.DataType
      Get
        Return _dataType
      End Get
    End Property

    ''' <summary>
    ''' Gets a value indicating whether this property
    ''' is an identity key for the object.
    ''' </summary>
    ''' <remarks>
    ''' Returns the optional value provided through
    ''' the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    ''' attribute on the property.
    ''' </remarks>
    Public ReadOnly Property Identity() As Boolean Implements System.Web.UI.Design.IDataSourceFieldSchema.Identity
      Get
        Return _isIdentity
      End Get
    End Property

    ''' <summary>
    ''' Gets a value indicating whether this property
    ''' is readonly.
    ''' </summary>
    Public ReadOnly Property IsReadOnly() As Boolean Implements System.Web.UI.Design.IDataSourceFieldSchema.IsReadOnly
      Get
        Return _isReadOnly
      End Get
    End Property

    ''' <summary>
    ''' Gets a value indicating whether this property
    ''' must contain a unique value.
    ''' </summary>
    ''' <returns>
    ''' Always returns True if the property
    ''' is marked as a primary key, otherwise
    ''' returns False.
    ''' </returns>
    Public ReadOnly Property IsUnique() As Boolean Implements System.Web.UI.Design.IDataSourceFieldSchema.IsUnique
      Get
        Return _primaryKey
      End Get
    End Property

    ''' <summary>
    ''' Gets the length of the property value.
    ''' </summary>
    ''' <remarks>
    ''' Returns the optional value provided through
    ''' the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    ''' attribute on the property.
    ''' </remarks>
    Public ReadOnly Property Length() As Integer Implements System.Web.UI.Design.IDataSourceFieldSchema.Length
      Get
        Return _length
      End Get
    End Property

    ''' <summary>
    ''' Gets the property name.
    ''' </summary>
    Public ReadOnly Property Name() As String _
      Implements System.Web.UI.Design.IDataSourceFieldSchema.Name
      Get
        Return _name
      End Get
    End Property

    ''' <summary>
    ''' Gets a value indicating whether the property
    ''' is nullable
    ''' </summary>
    ''' <remarks>
    ''' Returns True for reference types, and for
    ''' value types wrapped in the Nullable generic.
    ''' The result can also be set to True through
    ''' the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    ''' attribute on the property.
    ''' </remarks>
    Public ReadOnly Property Nullable() As Boolean _
      Implements System.Web.UI.Design.IDataSourceFieldSchema.Nullable
      Get
        Return _nullable
      End Get
    End Property

    ''' <summary>
    ''' Gets the property's numeric precision.
    ''' </summary>
    ''' <returns>Always returns -1.</returns>
    Public ReadOnly Property Precision() As Integer Implements System.Web.UI.Design.IDataSourceFieldSchema.Precision
      Get
        Return -1
      End Get
    End Property

    ''' <summary>
    ''' Gets a value indicating whether the property
    ''' is a primary key value.
    ''' </summary>
    ''' <remarks>
    ''' Returns the optional value provided through
    ''' the <see cref="DataObjectFieldAttribute">DataObjectField</see>
    ''' attribute on the property.
    ''' </remarks>
    Public ReadOnly Property PrimaryKey() As Boolean _
      Implements System.Web.UI.Design.IDataSourceFieldSchema.PrimaryKey
      Get
        Return _primaryKey
      End Get
    End Property

    ''' <summary>
    ''' Gets the property's scale.
    ''' </summary>
    ''' <returns>Always returns -1.</returns>
    Public ReadOnly Property Scale() As Integer _
      Implements System.Web.UI.Design.IDataSourceFieldSchema.Scale
      Get
        Return -1
      End Get
    End Property

  End Class

End Namespace